src/engine/SCons/Node/Python.py

```diff
@@ -133,7 +133,12 @@ def get_text_contents(self):
         ###TODO: something reasonable about universal newlines
         contents = str(self.value)
         for kid in self.children(None):
```

Comment:
Ugh, this feels like playing data type whack-a-mole: here we decode the result of `get_contents`, which if it's the `get_contents` just below, has just encoded it....

Reply:
indeed. well the idea is get_contents() always returns bytes so it can be hashed. get_text_contents() can be parsed (like for scanners). Though we should probably do something smart when it can't be Text'tified.

Reply:
I think the issue here shows up in certain cases with File node as well. I believe the problem is that in py2 it was just viewed as data ( we did not need to encode/decode it. meaning the data was good or bad from an encoding case). In python3 we have to say something about how to deal with the data. the current code assumes we are reading in text ( which is the common case for source files) however for many cases the source or most targets the "text" in binary data. The code needs to try to make text out of it, else fallback to it being a binary string.

Reply:
Okay, I get it.  The normal recommendation in these days of Py3 is to convert data at the edges - it's bytes coming from various I/O sources, then treat it as unicode all throughout, then goes back to bytes on the way out.  But the need to compute hashes, which want bytes, makes kind of a mess of that model.  So I'll be quiet now (though I still think there's too much converting going on).

Reply:
I guess I should ask the question does SCon need to have the content data in any form but byte strings?

Reply:
Not sure what you mean by "need"?
Do you mean something beyond what I explained above that get_content() should return bytes, get_text_content() should return a string.  (Of course in py27 both should be the same as strings are bytes there)

Reply:
This pull request also fixes a build error I ran into when upgrading VMware to use SCons 3.1.1 with Python 2. AFAICT the use case is if a Python Value has a file node as a child. Not sure if it's applicable in Python 3 too.

My question on this review is: why not just fix this by changing get_contents() to get_text_contents()? That function already manages a lot of the trial-and-error of retrieving file contents as a string, which is necessary in this function given that "contents" is already a string.

Reply:
@grossag -change it where?  as I said above. Hashes should be fed bytes. So get_contents() returns bytes. get_text_contents() returns str(). On py27 that of course is the same thing, but py3, it's not.

Reply:
@bdbaddog Change it in get_text_contents (the portion referenced in this comment thread in diff view). I'm just saying that get_text_contents function calls kid.get_contents().decode(), but it seems a lot easier to have get_text_contents call kid.get_text_contents() and leave all decoding and exception handling to the implementation of kid.get_text_contents.

Reply:
to be honest it would look like it should.. I think this code needs some clean up.

The main area of concern I have seen personally is that that the get_text_content has code like this:

`
contents = str(self.value)
        for kid in self.children(None):
            contents = contents + kid.get_contents().decode()
        return contents
`

which will add binary data to any children SCons might have on the value ( implict or explict). This is where I had added a PR ( I believe it was added in ... it seems to be part of your PR?) to not crash when the content is not unicode data. I have seen cases in which Values node get raw binary data files as depends.  The code above is very clear that the text content will possibly get binary data as it calls get_content() not get_text_content()

Reply:
As @dragon512 mentioned on the SCons Discord, another option it to do:

`contents = contents + kid.get_csig()`

That greatly helps with performance if a build uses a Value as a build target and then uses env.Depends() to have that Value depend on several nodes.

Reply:
To be honest. I'm not sure why Value() node is including the contents or csig of it's children when asked for get_text_contents(). To me that make zero sense.